src/Components/Components/src/ComponentFactory.cs

```diff
@@ -14,46 +15,76 @@ internal sealed class ComponentFactory
     private const BindingFlags _injectablePropertyBindingFlags
         = BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic;
 
-    private static readonly ConcurrentDictionary<Type, Action<IServiceProvider, IComponent>> _cachedInitializers = new();
+    private static readonly ConcurrentDictionary<Type, (Action<IServiceProvider, IComponent> propertyInitializers, ObjectFactory? componentInitializer)> _cachedInitializers = new();
+    private readonly IComponentActivator? _componentActivator;
 
-    private readonly IComponentActivator _componentActivator;
-
-    public ComponentFactory(IComponentActivator componentActivator)
+    public ComponentFactory(IComponentActivator? componentActivator)
     {
-        _componentActivator = componentActivator ?? throw new ArgumentNullException(nameof(componentActivator));
+        _componentActivator = componentActivator;
     }
 
     public static void ClearCache() => _cachedInitializers.Clear();
 
     public IComponent InstantiateComponent(IServiceProvider serviceProvider, [DynamicallyAccessedMembers(Component)] Type componentType)
     {
-        var component = _componentActivator.CreateInstance(componentType);
-        if (component is null)
+        if (_componentActivator is not null)
         {
-            // The default activator will never do this, but an externally-supplied one might
-            throw new InvalidOperationException($"The component activator returned a null value for a component of type {componentType.FullName}.");
+            return InstantiateWithActivator(_componentActivator, serviceProvider, componentType);
         }
 
-        PerformPropertyInjection(serviceProvider, component);
+        return InstantiateDefault(serviceProvider, componentType);
+    }
+
+    private static IComponent InstantiateDefault(IServiceProvider serviceProvider, [DynamicallyAccessedMembers(Component)] Type componentType)
+    {
+        // This is thread-safe because _cachedInitializers is a ConcurrentDictionary.
+        // We might generate the initializer more than once for a given type, but would
+        // still produce the correct result.
```

Comment:
Any reason not to use `ConcurrentDictionary`'s `GetOrAdd` method here?

Reply:
It's a bit easier to avoid the extra static delegate allocation which this pattern. I also worry that the static delegate is very easy to turn in to a per invocation allocation if we aren't careful about how it gets used.